apps/server/src/modules/authorization/authorization.helper.ts

```diff
@@ -0,0 +1,97 @@
+import { Collection } from '@mikro-orm/core';
+import { Injectable } from '@nestjs/common';
+import { Role, User } from '@shared/domain/entity';
+
+@Injectable()
+export class AuthorizationHelper {
+	/**
+	 * Recursively resolve all roles and permissions of a user.
+	 * IMPORTANT: The role collections of the user and nested roles will not be loaded lazily.
+	 * Please make sure you populate them before calling this method.
+	 */
+	public resolvePermissions(user: User): string[] {
+		if (!user.roles.isInitialized(true)) {
+			throw new Error('Roles items are not loaded.');
+		}
+		const rolesAndPermissions = this.resolvePermissionsByRoles(user.roles.getItems());
+
+		return rolesAndPermissions;
+	}
+
+	private resolvePermissionsByRoles(inputRoles: Role[]): string[] {
+		let permissions: string[] = [];
+
+		for (let i = 0; i < inputRoles.length; i += 1) {
+			const role = inputRoles[i];
+			if (!role.roles.isInitialized(true)) {
+				throw new Error('Roles items are not loaded.');
+			}
+			const innerRoles = role.roles.getItems();
+			permissions = [...permissions, ...role.permissions];
+
+			if (innerRoles.length > 0) {
+				const subPermissions = this.resolvePermissionsByRoles(innerRoles);
+				permissions = [...permissions, ...subPermissions];
+			}
+		}
```

Comment:
```suggestion
		for (let i = 0; i < inputRoles.length; i += 1) {
			const role = inputRoles[i];
			if (!role.roles.isInitialized(true)) {
				throw new Error('Roles items are not loaded.');
			}
			permissions = [...permissions, ...role.permissions];
                        
                        const innerRoles = role.roles.getItems();
			if (innerRoles.length > 0) {
				const subPermissions = this.resolvePermissionsByRoles(innerRoles);
				permissions = [...permissions, ...subPermissions];
			}
		}
```
Also thought about breaking it down into smaller functions. But thats giving me a hard time somehow. Maybe like this?
```suggestion
		for (let i = 0; i < inputRoles.length; i += 1) {
			const rolePermissions = getRolePermissions(inputRoles[i]);
			permissions = [...permissions, ...rolePermissions];
                        
                        const innerRoles = getInnerRolePermissions(inputRoles[i]);
			permissions = [...permissions, ...innerRoles];
			
		}
```